### ISSUE TITLE
DNS cookies

### ISSUE THREAD

**Issue by jpds** [hickory-dns/hickory-dns#2075]
**Describe the solution you'd like**

Server and client-side support for DNS Cookies:

- https://datatracker.ietf.org/doc/html/rfc7873
- https://datatracker.ietf.org/doc/html/rfc9018

More information at [this video from CIRA](https://www.youtube.com/watch?v=t_VZP4s5q5c) (3:24 onwards)

**Comment by bluejekyll** [comment]
Reading through this, it should be fairly straightforward to implement. Getting the client and server cookie creation correct looks like the most complicated part.

The only question I have is, given that we have encrypted options available with TLS, h2, h3, and quic, what is the priority of this RFC?

**Comment by djc** [comment]
Do we enable DoT/DoH/DoH3/DoQ by default, though, or do most configurations in the wild still end up with UDP/TCP?

**Comment by bluejekyll** [comment]
I’m just trying to an understanding of how big the community of cookie users there are.

overall, this seems somewhat straightforward to implement, but my point was that it’s a bit redundant to these other options is all.

**Comment by bluejekyll** [comment]
Here's some good discussion into the usage of Cookies, https://blog.apnic.net/2023/02/23/notes-from-dns-oarc-40/, and a presentation from Google on the same: https://indico.dns-oarc.net/event/46/contributions/978/attachments/947/1754/Cache%20Poisoning%20Protection%20-%20Deployment%20Experience.pdf

**Comment by bluejekyll** [comment]
Yup, I don't think this is a problem to implement. Are you interested in taking a stab at it? I think it could be broken up into two PRs. One to add the parsing to the `EdnsOption` enum:

https://github.com/hickory-dns/hickory-dns/blob/bb203245629f4cc4e7a48861b7d6810c011393c7/crates/proto/src/rr/rdata/opt.rs#L447C10-L447C10

And then another to put this logic into the resolver and server implementations. This one will be the more interesting logic to write... I'm guessing that we only want it on the UDP implementations. The server side logic is perhaps simpler, as the Client needs to determine if the server supports Cookies or not. If it doesn't, then it needs to drop the Cookies from future requests... And based on the above links, it looks like there are some servers out there that fail in the presence of Cookies.
